### PR TITLE
RavenDB-19762: Expanding debug information for cases where adding a term for a field failed

### PR DESCRIPTION
**Pull request by ArieSLV**
There was a situation when we could not determine the cause of the exception due to insufficient data specified in the exception message. To address this in future cases, a handler for `IndexOutOfRangeException` has been added that provides information about the current state of the term whenever the exception occurs.

### CODE REVIEW

File: src/Lucene.Net/Index/TermsHashPerField.cs [hunk 1]
```diff
@@ -355,168 +355,175 @@ public void  Add(int textStart)
...
+                    intUptoStart = p.intStart & DocumentsWriter.INT_BLOCK_MASK;
+                    consumer.AddTerm(p);
+                }
+
+                if (doNextCall)
+                    nextPerField.Add(p.textStart);
+            }
+            catch (IndexOutOfRangeException e)
+            {
+                throw new IndexOutOfRangeException($"Failed to add term for field '{fieldInfo.name}', term: '{termAtt.Term}', tokenText: '{new string(tokenText)}', tokenTextLength: '{tokenTextLen}'.", e);
```
Comment (grisha-kotler) [inline src/Lucene.Net/Index/TermsHashPerField.cs]: What happens if `tokenText` is null?

**Review by grisha-kotler — APPROVED** [review]

File: src/Lucene.Net/Index/TermsHashPerField.cs [hunk 2]
```diff
@@ -355,168 +355,175 @@ public void  Add(int textStart)
...
+                {
+                    intUptos = intPool.buffers[p.intStart >> DocumentsWriter.INT_BLOCK_SHIFT];
+                    intUptoStart = p.intStart & DocumentsWriter.INT_BLOCK_MASK;
+                    consumer.AddTerm(p);
+                }
+
+                if (doNextCall)
+                    nextPerField.Add(p.textStart);
+            }
+            catch (IndexOutOfRangeException e)
```
Comment (grisha-kotler) [inline src/Lucene.Net/Index/TermsHashPerField.cs]: And let's catch any exception here.

File: src/Lucene.Net/Index/TermsHashPerField.cs (hunk 1 above)
Comment (ArieSLV) [reply src/Lucene.Net/Index/TermsHashPerField.cs]: `string.Empty`

File: src/Lucene.Net/Index/TermsHashPerField.cs (hunk 2 above)
Comment (ArieSLV) [reply src/Lucene.Net/Index/TermsHashPerField.cs]: Done.

**Review by grisha-kotler — APPROVED** [review]

File: src/Lucene.Net/Index/TermsHashPerField.cs [hunk 3]
```diff
@@ -355,168 +355,176 @@ public void  Add(int textStart)
...
+                    consumer.AddTerm(p);
+                }
+
+                if (doNextCall)
+                    nextPerField.Add(p.textStart);
+            }
+            catch (Exception e)
+            {
+                var field = docState.doc.GetField("id()");
+                throw new InvalidOperationException($"Failed to add term for document field '{fieldInfo.name}', docId: '{field?.fieldsData}', term: '{termAtt.Term}', tokenText: '{new string(tokenText)}', tokenTextLength: '{tokenTextLen}'.", e);
```
Comment (ppekrol) [inline src/Lucene.Net/Index/TermsHashPerField.cs]: hard for me to follow changes here, except of wrapping everything in try/catch are there any other changes?
I'm a bit worried about wrapping each exception in 'InvalidOperationException' - is there a specific exception we want to catch? Are there any expected exception that we shouldn't catch?

Comment (ArieSLV) [reply src/Lucene.Net/Index/TermsHashPerField.cs]: A refactoring was carried out, in which:
1. Error handling moved to caller level.
2. Only the error that prompted us to make these changes is processed (`IndexOutOfRangeException`)

The information content of the error message is preserved.
An example of a previous implementation error:
`Failed to add term for document field 'Name', docId: 'contract/1-A', term: 'quotes/98ed381e-116a-46c4-b25a-6c7d4e89c239'`

Now the message looks like this:
`Failed to add term for document field 'Name', docId: 'contract/1-A', field attributes: (startOffset=0,endOffset=43,term=quotes/98ed381e-116a-46c4-b25a-6c7d4e89c239 ,positionIncrement=1)`

**Review by grisha-kotler — APPROVED** [review]

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**